### PR TITLE
TS-29104 Initial implementation of allowing multiple include and excl…

### PR DESCRIPTION
**Pull request by j-christl**
…ude patterns

Addresses issue [TS-29104](https://cqse.atlassian.net/browse/TS-29104)

- [ ] Changes are tested adequately
- [ ] Profilers's README.md and the Teamscale documentation are updated in case of user-visible changes
- [ ] CHANGELOG.md of the affected packages are updated
- [ ] Present new features in [N&N](https://wiki.cqse.eu/pages/viewpage.action?pageId=689566)

Please respect the vote of the [Teamscale bot](https://demo.teamscale.com) or flag irrelevant findings as tolerated or false positives. If you feel that the Teamscale config needs adjustment, please state so in a comment and discuss this with your reviewer.

### CODE REVIEW

File: test/casestudies/angular-hero-app/package-lock.json [hunk 1]
```diff
@@ -15186,7 +15186,9 @@
       "resolved": "https://registry.npmjs.org/ajv-formats/-/ajv-formats-2.1.1.tgz",
       "integrity": "sha512-Wx0Kx52hxE7C18hkMEggYlEifqWZtYaRgouJor+WMdPnQyEK13vgEWyVNup7SoeeoLMsr4kf5h6dOW11I15MUA==",
       "dev": true,
-      "requires": {}
+      "requires": {
+        "ajv": "^8.0.0"
```
Comment (j-christl) [inline test/casestudies/angular-hero-app/package-lock.json:15190]: This change should not be committed right?

File: packages/teamscale-javascript-instrumenter/src/instrumenter/Task.ts [hunk 2]
```diff
@@ -65,17 +65,35 @@ export class CollectorSpecifier {
...
-	private readonly exclude: string | undefined;
-
-	constructor(include: string | undefined, exclude: string | undefined) {
-		this.include = OriginSourcePattern.normalizeGlobPattern(include);
-		this.exclude = OriginSourcePattern.normalizeGlobPattern(exclude);
+	private readonly exclude: string[] | undefined;
+
+	constructor(include: string[] | undefined, exclude: string[] | undefined) {
+		this.include = [];
+		for (const includeEntry of include || []) {
```
Comment (stahlbauer) [inline packages/teamscale-javascript-instrumenter/src/instrumenter/Task.ts:78]: The control flow seems to be quite complicate for the functionality it implements.

Would it be possible to replace this loop by a map on the array (if defined). Also `normalizeGlobPattern` could be adjusted to only work for defined values.

File: packages/teamscale-javascript-instrumenter/src/instrumenter/Task.ts [hunk 3]
```diff
@@ -65,17 +65,35 @@ export class CollectorSpecifier {
...
+			const normalizedIncludeEntry = OriginSourcePattern.normalizeGlobPattern(includeEntry);
+			if (normalizedIncludeEntry !== undefined) {
+				this.include?.push(<string>normalizedIncludeEntry);
+			}
+		}
+		if (this.include.length === 0) {
+			this.include = undefined;
+		}
+		this.exclude = [];
+		for (const excludeEntry of exclude || []) {
```
Comment (stahlbauer) [inline packages/teamscale-javascript-instrumenter/src/instrumenter/Task.ts:88]: The same as for the `includes` applies here. Rising the question if this isn't a code clone. Separate function for normalizing a given array?

File: packages/teamscale-javascript-instrumenter/src/instrumenter/Task.ts (hunk 2 above)
Comment (j-christl) [reply packages/teamscale-javascript-instrumenter/src/instrumenter/Task.ts]: Thanks! Introduced normalizePatterns now to apply the same logic to include and exclude patterns. Also simplified for loop

File: test/systemtest.mjs [hunk 4]
```diff
@@ -12,39 +12,61 @@ import ServerMock from 'mock-http-server';
...
+		expectUncoveredLines: {},
+		excludeOrigins: [],
+		includeOrigins: []
 	},
-	{name: 'vite-react-app',
+	{
+		name: 'vite-react-app',
 		rootDir: 'test/casestudies/vite-react-app',
-		includePattern: "../../src/**/*.*",
+		includePattern: '../../src/**/*.*',
```
Comment (stahlbauer) [inline test/systemtest.mjs:30]: Could you use this as an `includeOrigins` pattern? (also for the other case studies).

Comment (j-christl) [reply test/systemtest.mjs]: Thanks for noticing! Fixed this now. Also fixed a bug in expectCoveredLines, probably caused by the merge

**Pull request merged**